src/Lean/Elab/Tactic/BuiltinTactic.lean

```diff
@@ -35,55 +35,52 @@ open Parser.Tactic
 @[builtinTactic paren] def evalParen : Tactic := fun stx =>
   evalTactic stx[1]
 
-def isCheckpointableTactic (arg : Syntax) : TacticM Bool := do
+def isCheckpointableTactic (arg : Syntax.Tactic) : TacticM Bool := do
   -- TODO: make it parametric
-  let kind := arg.getKind
+  let kind := arg.1.getKind
   return kind == ``Lean.Parser.Tactic.save
 
-partial def addCheckpoints (args : Array Syntax) : TacticM (Array Syntax) := do
+partial def addCheckpoints (args : Array Syntax.Tactic) : TacticM (Array Syntax.Tactic) := do
   -- if (← readThe Term.Context).tacticCache? |>.isSome then
-  if (← args.anyM fun arg => isCheckpointableTactic arg[0]) then
+  if (← args.anyM fun arg => isCheckpointableTactic arg) then
     let argsNew ← go 0 #[] #[]
     return argsNew
   return args
 where
-  push (acc : Array Syntax) (result : Array Syntax) : Array Syntax :=
+  push (acc : Array Syntax.Tactic) (result : Array Syntax.Tactic) : Array Syntax.Tactic :=
     if acc.isEmpty then
       result
     else
-      let ref := acc.back
-      let accSeq := mkNode ``Lean.Parser.Tactic.tacticSeq1Indented #[mkNullNode acc]
-      let checkpoint := mkNode ``Lean.Parser.Tactic.checkpoint #[mkAtomFrom ref "checkpoint", accSeq]
-      result.push (mkNode groupKind #[checkpoint, mkNullNode])
+      result.push <| Unhygienic.run do withRef acc.back do
+        `(tactic| checkpoint $[$acc]*)
 
-  go (i : Nat) (acc : Array Syntax) (result : Array Syntax) : TacticM (Array Syntax) := do
+  go (i : Nat) (acc : Array Syntax.Tactic) (result : Array Syntax.Tactic) : TacticM (Array Syntax.Tactic) := do
     if h : i < args.size then
       let arg := args.get ⟨i, h⟩
-      if (← isCheckpointableTactic arg[0]) then
+      if (← isCheckpointableTactic arg) then
         -- `argNew` is `arg` as singleton sequence. The idea is to make sure it does not satisfy `isCheckpointableTactic` anymore
-        let argNew := arg.setArg 0 (mkNullNode #[arg[0]])
+        let argNew := ⟨mkNullNode #[arg]⟩ -- HACK
```

Comment:
Could you please explain the "HACK"?

Reply:
It's a hack because it creates invalid syntax (see pretty-printer issues in the Zulip thread):

```
import Lean open Lean Elab Command PrettyPrinter

elab "run" : command => do
  let tac ← `(tactic| simp)
  let tac : TSyntax `tactic := ⟨mkNullNode #[tac]⟩ -- HACK: produces invalid tactic
  let term ← `(by $tac:tactic)
  logInfo (← liftCoreM do ppTerm term)

run -- unknown constant 'null'
```